### ISSUE TITLE
Create shared openai-compatible inference adapter

### ISSUE THREAD

**Issue by russellb** [meta-llama/llama-stack#193]
There are multiple open PRs looking to add inference adapters for services that offer OpenAI-compatible APIs.

* databricks - https://github.com/meta-llama/llama-stack/pull/83
* sambanova - https://github.com/meta-llama/llama-stack/pull/136
* runpod - https://github.com/meta-llama/llama-stack/pull/157
* vllm - https://github.com/meta-llama/llama-stack/pull/178

These adapters all use the OpenAI client and have the same config options (a URL and API key).

Options for moving forward:

1. ~~Create a single shared adapter for OpenAI-compatible API endpoints.~~
2. ~~If some cases require additional unique configuration, most of the code could be shared.~~
3. Create code to be shared by each adapter that uses an OpenAI-compatible API.

**Comment by ashwinb** [comment]
I support creating a shared code but not a shared adapter. I think implementation quality in terms of faithfulness of precise token-level behavior (on the backend) varies so it is best if we keep a top-level fork but share underlying code where it makes sense. I also plan to put together a set of tests which test tool calling behavior and other corner-case behavior we know can be problematic.

Please do hold off on doing this work for a couple days. There's some impending refactors incoming. I will be merging some PRs in the meantime to avoid duplicate work on behalf of the providers.

**Comment by terrytangyuan** [comment]
@ashwinb I have put together some work on this before noticing this issue and your comment: https://github.com/meta-llama/llama-stack/pull/195

Is it possible to merge that effort and then move them to reusable code in a separate PR (or as part of your incoming impending refactors) if a shared adapter is not preferred? This will unblock some of our existing tasks and avoid duplicating work.

**Comment by nathan-weinberg** [comment]
Is this issue still relevant? All the linked PRs seem to have been merged or closed
